File: csharp/src/main/csharp/IO/Swagger/Model/PresenceInfo.cs
using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace IO.Swagger.Model
{
    /// <summary>
    /// 
    /// </summary>
    [DataContract]
    public partial class PresenceInfo :  IEquatable<PresenceInfo>
    { 
    
        /// <summary>
        /// Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The 'DoNotAcceptDepartmentCalls' and 'TakeDepartmentCallsOnly' values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The 'TakeDepartmentCallsOnly' status can be set through the old RingCentral user interface and is available for some migrated accounts only.
        /// </summary>
        /// <value>Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The 'DoNotAcceptDepartmentCalls' and 'TakeDepartmentCallsOnly' values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The 'TakeDepartmentCallsOnly' status can be set through the old RingCentral user interface and is available for some migrated accounts only.</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum DndStatusEnum {
            
            [EnumMember(Value = "TakeAllCalls")]
            Takeallcalls,
            
            [EnumMember(Value = "DoNotAcceptAnyCalls")]
            Donotacceptanycalls,
            
            [EnumMember(Value = "DoNotAcceptDepartmentCalls")]
            Donotacceptdepartmentcalls,
            
            [EnumMember(Value = "TakeDepartmentCallsOnly")]
            Takedepartmentcallsonly
        }

    
        /// <summary>
        /// Aggregated presence status, calculated from a number of sources
        /// </summary>
        /// <value>Aggregated presence status, calculated from a number of sources</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum PresenceStatusEnum {
            
            [EnumMember(Value = "Offline")]
            Offline,
            
            [EnumMember(Value = "Busy")]
            Busy,
            
            [EnumMember(Value = "Available")]
            Available
        }

    
        /// <summary>
        /// Telephony presence status
        /// </summary>
        /// <value>Telephony presence status</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum TelephonyStatusEnum {
            
            [EnumMember(Value = "NoCall")]
            Nocall,
            
            [EnumMember(Value = "CallConnected")]
            Callconnected,
            
            [EnumMember(Value = "Ringing")]
            Ringing,
            
            [EnumMember(Value = "OnHold")]
            Onhold,
            
            [EnumMember(Value = "ParkedCall")]
            Parkedcall
        }

    
        /// <summary>
        /// User-defined presence status (as previously published by the user)
        /// </summary>
        /// <value>User-defined presence status (as previously published by the user)</value>
        [JsonConverter(typeof(StringEnumConverter))]
        public enum UserStatusEnum {
            
            [EnumMember(Value = "Offline")]
            Offline,
            
            [EnumMember(Value = "Busy")]
            Busy,
            
            [EnumMember(Value = "Available")]
            Available
        }

    
        /// <summary>
        /// Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The 'DoNotAcceptDepartmentCalls' and 'TakeDepartmentCallsOnly' values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The 'TakeDepartmentCallsOnly' status can be set through the old RingCentral user interface and is available for some migrated accounts only.
        /// </summary>
        /// <value>Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The 'DoNotAcceptDepartmentCalls' and 'TakeDepartmentCallsOnly' values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The 'TakeDepartmentCallsOnly' status can be set through the old RingCentral user interface and is available for some migrated accounts only.</value>
        [DataMember(Name="dndStatus", EmitDefaultValue=false)]
        public DndStatusEnum? DndStatus { get; set; }
    
        /// <summary>
        /// Aggregated presence status, calculated from a number of sources
        /// </summary>
        /// <value>Aggregated presence status, calculated from a number of sources</value>
        [DataMember(Name="presenceStatus", EmitDefaultValue=false)]
        public PresenceStatusEnum? PresenceStatus { get; set; }
    
        /// <summary>
        /// Telephony presence status
        /// </summary>
        /// <value>Telephony presence status</value>
        [DataMember(Name="telephonyStatus", EmitDefaultValue=false)]
        public TelephonyStatusEnum? TelephonyStatus { get; set; }
    
        /// <summary>
        /// User-defined presence status (as previously published by the user)
        /// </summary>
        /// <value>User-defined presence status (as previously published by the user)</value>
        [DataMember(Name="userStatus", EmitDefaultValue=false)]
        public UserStatusEnum? UserStatus { get; set; }
    
        /// <summary>
        /// Initializes a new instance of the <see cref="PresenceInfo" /> class.
        /// Initializes a new instance of the <see cref="PresenceInfo" />class.
        /// </summary>
        /// <param name="Uri">Canonical URI of a presence info resource.</param>
        /// <param name="AllowSeeMyPresence">If &#39;True&#39; enables other extensions to see the extension presence status.</param>
        /// <param name="DndStatus">Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The &#39;DoNotAcceptDepartmentCalls&#39; and &#39;TakeDepartmentCallsOnly&#39; values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The &#39;TakeDepartmentCallsOnly&#39; status can be set through the old RingCentral user interface and is available for some migrated accounts only..</param>
        /// <param name="Extension">Extension.</param>
        /// <param name="Message">Custom status message (as previously published by user).</param>
        /// <param name="PickUpCallsOnHold">If &#39;True&#39; enables the extension user to pick up a monitored line on hold.</param>
        /// <param name="PresenceStatus">Aggregated presence status, calculated from a number of sources.</param>
        /// <param name="RingOnMonitoredCall">If &#39;True&#39; enables to ring extension phone, if any user monitored by this extension is ringing.</param>
        /// <param name="TelephonyStatus">Telephony presence status.</param>
        /// <param name="UserStatus">User-defined presence status (as previously published by the user).</param>

        public PresenceInfo(string Uri = null, bool? AllowSeeMyPresence = null, DndStatusEnum? DndStatus = null, PresenceInfoExtensionInfo Extension = null, string Message = null, bool? PickUpCallsOnHold = null, PresenceStatusEnum? PresenceStatus = null, bool? RingOnMonitoredCall = null, TelephonyStatusEnum? TelephonyStatus = null, UserStatusEnum? UserStatus = null)
        {
            this.Uri = Uri;
            this.AllowSeeMyPresence = AllowSeeMyPresence;
            this.DndStatus = DndStatus;
            this.Extension = Extension;
            this.Message = Message;
            this.PickUpCallsOnHold = PickUpCallsOnHold;
            this.PresenceStatus = PresenceStatus;
            this.RingOnMonitoredCall = RingOnMonitoredCall;
            this.TelephonyStatus = TelephonyStatus;
            this.UserStatus = UserStatus;
            
        }
        
    
        /// <summary>
        /// Canonical URI of a presence info resource
        /// </summary>
        /// <value>Canonical URI of a presence info resource</value>
        [DataMember(Name="uri", EmitDefaultValue=false)]
        public string Uri { get; set; }
    
        /// <summary>
        /// If &#39;True&#39; enables other extensions to see the extension presence status
        /// </summary>
        /// <value>If &#39;True&#39; enables other extensions to see the extension presence status</value>
        [DataMember(Name="allowSeeMyPresence", EmitDefaultValue=false)]
        public bool? AllowSeeMyPresence { get; set; }
    
        /// <summary>
        /// Gets or Sets Extension
        /// </summary>
        [DataMember(Name="extension", EmitDefaultValue=false)]
        public PresenceInfoExtensionInfo Extension { get; set; }
    
        /// <summary>
        /// Custom status message (as previously published by user)
        /// </summary>
        /// <value>Custom status message (as previously published by user)</value>
        [DataMember(Name="message", EmitDefaultValue=false)]
        public string Message { get; set; }
    
        /// <summary>
        /// If &#39;True&#39; enables the extension user to pick up a monitored line on hold
        /// </summary>
        /// <value>If &#39;True&#39; enables the extension user to pick up a monitored line on hold</value>
        [DataMember(Name="pickUpCallsOnHold", EmitDefaultValue=false)]
        public bool? PickUpCallsOnHold { get; set; }
    
        /// <summary>
        /// If &#39;True&#39; enables to ring extension phone, if any user monitored by this extension is ringing
        /// </summary>
        /// <value>If &#39;True&#39; enables to ring extension phone, if any user monitored by this extension is ringing</value>
        [DataMember(Name="ringOnMonitoredCall", EmitDefaultValue=false)]
        public bool? RingOnMonitoredCall { get; set; }
    
        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.Append("class PresenceInfo {\n");
            sb.Append("  Uri: ").Append(Uri).Append("\n");
            sb.Append("  AllowSeeMyPresence: ").Append(AllowSeeMyPresence).Append("\n");
            sb.Append("  DndStatus: ").Append(DndStatus).Append("\n");
            sb.Append("  Extension: ").Append(Extension).Append("\n");
            sb.Append("  Message: ").Append(Message).Append("\n");
            sb.Append("  PickUpCallsOnHold: ").Append(PickUpCallsOnHold).Append("\n");
            sb.Append("  PresenceStatus: ").Append(PresenceStatus).Append("\n");
            sb.Append("  RingOnMonitoredCall: ").Append(RingOnMonitoredCall).Append("\n");
            sb.Append("  TelephonyStatus: ").Append(TelephonyStatus).Append("\n");
            sb.Append("  UserStatus: ").Append(UserStatus).Append("\n");
            
            sb.Append("}\n");
            return sb.ToString();
        }
  
        /// <summary>
        /// Returns the JSON string presentation of the object
        /// </summary>
        /// <returns>JSON string presentation of the object</returns>
        public string ToJson()
        {
            return JsonConvert.SerializeObject(this, Formatting.Indented);
        }

        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="obj">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object obj)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            return this.Equals(obj as PresenceInfo);
        }

        /// <summary>
        /// Returns true if PresenceInfo instances are equal
        /// </summary>
        /// <param name="other">Instance of PresenceInfo to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(PresenceInfo other)
        {
            // credit: http://stackoverflow.com/a/10454552/677735
            if (other == null)
                return false;

            return 
                (
                    this.Uri == other.Uri ||
                    this.Uri != null &&
                    this.Uri.Equals(other.Uri)
                ) && 
                (
                    this.AllowSeeMyPresence == other.AllowSeeMyPresence ||
                    this.AllowSeeMyPresence != null &&
                    this.AllowSeeMyPresence.Equals(other.AllowSeeMyPresence)
                ) && 
                (
                    this.DndStatus == other.DndStatus ||
                    this.DndStatus != null &&
                    this.DndStatus.Equals(other.DndStatus)
                ) && 
                (
                    this.Extension == other.Extension ||
                    this.Extension != null &&
                    this.Extension.Equals(other.Extension)
                ) && 
                (
                    this.Message == other.Message ||
                    this.Message != null &&
                    this.Message.Equals(other.Message)
                ) && 
                (
                    this.PickUpCallsOnHold == other.PickUpCallsOnHold ||
                    this.PickUpCallsOnHold != null &&
                    this.PickUpCallsOnHold.Equals(other.PickUpCallsOnHold)
                ) && 
                (
                    this.PresenceStatus == other.PresenceStatus ||
                    this.PresenceStatus != null &&
                    this.PresenceStatus.Equals(other.PresenceStatus)
                ) && 
                (
                    this.RingOnMonitoredCall == other.RingOnMonitoredCall ||
                    this.RingOnMonitoredCall != null &&
                    this.RingOnMonitoredCall.Equals(other.RingOnMonitoredCall)
                ) && 
                (
                    this.TelephonyStatus == other.TelephonyStatus ||
                    this.TelephonyStatus != null &&
                    this.TelephonyStatus.Equals(other.TelephonyStatus)
                ) && 
                (
                    this.UserStatus == other.UserStatus ||
                    this.UserStatus != null &&
                    this.UserStatus.Equals(other.UserStatus)
                );
        }

        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            // credit: http://stackoverflow.com/a/263416/677735
            unchecked // Overflow is fine, just wrap
            {
                int hash = 41;
                // Suitable nullity checks etc, of course :)
                
                if (this.Uri != null)
                    hash = hash * 59 + this.Uri.GetHashCode();
                
                if (this.AllowSeeMyPresence != null)
                    hash = hash * 59 + this.AllowSeeMyPresence.GetHashCode();
                
                if (this.DndStatus != null)
                    hash = hash * 59 + this.DndStatus.GetHashCode();
                
                if (this.Extension != null)
                    hash = hash * 59 + this.Extension.GetHashCode();
                
                if (this.Message != null)
                    hash = hash * 59 + this.Message.GetHashCode();
                
                if (this.PickUpCallsOnHold != null)
                    hash = hash * 59 + this.PickUpCallsOnHold.GetHashCode();
                
                if (this.PresenceStatus != null)
                    hash = hash * 59 + this.PresenceStatus.GetHashCode();
                
                if (this.RingOnMonitoredCall != null)
                    hash = hash * 59 + this.RingOnMonitoredCall.GetHashCode();
                
                if (this.TelephonyStatus != null)
                    hash = hash * 59 + this.TelephonyStatus.GetHashCode();
                
                if (this.UserStatus != null)
                    hash = hash * 59 + this.UserStatus.GetHashCode();
                
                return hash;
            }
        }

    }
}

File: javascript/src/model/PresenceInfo.js
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['../ApiClient', './PresenceInfoExtensionInfo'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'), require('./PresenceInfoExtensionInfo'));
  } else {
    // Browser globals (root is window)
    if (!root.RingCentralApi) {
      root.RingCentralApi = {};
    }
    root.RingCentralApi.PresenceInfo = factory(root.RingCentralApi.ApiClient, root.RingCentralApi.PresenceInfoExtensionInfo);
  }
}(this, function(ApiClient, PresenceInfoExtensionInfo) {
  'use strict';

  /**
   * The PresenceInfo model module.
   * @module model/PresenceInfo
   * @version v1.0
   */

  /**
   * Constructs a new <code>PresenceInfo</code>.
   * @alias module:model/PresenceInfo
   * @class
   */
  var exports = function() {











  };

  /**
   * Constructs a <code>PresenceInfo</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/PresenceInfo} obj Optional instance to populate.
   * @return {module:model/PresenceInfo} The populated <code>PresenceInfo</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) { 
      obj = obj || new exports();

      if (data.hasOwnProperty('uri')) {
        obj['uri'] = ApiClient.convertToType(data['uri'], 'String');
      }
      if (data.hasOwnProperty('allowSeeMyPresence')) {
        obj['allowSeeMyPresence'] = ApiClient.convertToType(data['allowSeeMyPresence'], 'Boolean');
      }
      if (data.hasOwnProperty('dndStatus')) {
        obj['dndStatus'] = ApiClient.convertToType(data['dndStatus'], 'String');
      }
      if (data.hasOwnProperty('extension')) {
        obj['extension'] = PresenceInfoExtensionInfo.constructFromObject(data['extension']);
      }
      if (data.hasOwnProperty('message')) {
        obj['message'] = ApiClient.convertToType(data['message'], 'String');
      }
      if (data.hasOwnProperty('pickUpCallsOnHold')) {
        obj['pickUpCallsOnHold'] = ApiClient.convertToType(data['pickUpCallsOnHold'], 'Boolean');
      }
      if (data.hasOwnProperty('presenceStatus')) {
        obj['presenceStatus'] = ApiClient.convertToType(data['presenceStatus'], 'String');
      }
      if (data.hasOwnProperty('ringOnMonitoredCall')) {
        obj['ringOnMonitoredCall'] = ApiClient.convertToType(data['ringOnMonitoredCall'], 'Boolean');
      }
      if (data.hasOwnProperty('telephonyStatus')) {
        obj['telephonyStatus'] = ApiClient.convertToType(data['telephonyStatus'], 'String');
      }
      if (data.hasOwnProperty('userStatus')) {
        obj['userStatus'] = ApiClient.convertToType(data['userStatus'], 'String');
      }
    }
    return obj;
  }


  /**
   * Canonical URI of a presence info resource
   * @member {String} uri
   */
  exports.prototype['uri'] = undefined;

  /**
   * If 'True' enables other extensions to see the extension presence status
   * @member {Boolean} allowSeeMyPresence
   */
  exports.prototype['allowSeeMyPresence'] = undefined;

  /**
   * Extended DnD (Do not Disturb) status. Cannot be set for Department/Announcement/Voicemail (Take Messages Only)/Fax User/Shared Lines Group/Paging Only Group/IVR Menu/Application Extension/Park Location extensions. The 'DoNotAcceptDepartmentCalls' and 'TakeDepartmentCallsOnly' values are applicable only for extensions - members of a Department; if these values are set for department outsiders, the 400 Bad Request error code is returned. The 'TakeDepartmentCallsOnly' status can be set through the old RingCentral user interface and is available for some migrated accounts only.
   * @member {module:model/PresenceInfo.DndStatusEnum} dndStatus
   */
  exports.prototype['dndStatus'] = undefined;

  /**
   * @member {module:model/PresenceInfoExtensionInfo} extension
   */
  exports.prototype['extension'] = undefined;

  /**
   * Custom status message (as previously published by user)
   * @member {String} message
   */
  exports.prototype['message'] = undefined;

  /**
   * If 'True' enables the extension user to pick up a monitored line on hold
   * @member {Boolean} pickUpCallsOnHold
   */
  exports.prototype['pickUpCallsOnHold'] = undefined;

  /**
   * Aggregated presence status, calculated from a number of sources
   * @member {module:model/PresenceInfo.PresenceStatusEnum} presenceStatus
   */
  exports.prototype['presenceStatus'] = undefined;

  /**
   * If 'True' enables to ring extension phone, if any user monitored by this extension is ringing
   * @member {Boolean} ringOnMonitoredCall
   */
  exports.prototype['ringOnMonitoredCall'] = undefined;

  /**
   * Telephony presence status
   * @member {module:model/PresenceInfo.TelephonyStatusEnum} telephonyStatus
   */
  exports.prototype['telephonyStatus'] = undefined;

  /**
   * User-defined presence status (as previously published by the user)
   * @member {module:model/PresenceInfo.UserStatusEnum} userStatus
   */
  exports.prototype['userStatus'] = undefined;


  /**
   * Allowed values for the <code>dndStatus</code> property.
   * @enum {String}
   * @readonly
   */
  exports.DndStatusEnum = { 
    /**
     * value: TakeAllCalls
     * @const
     */
    TAKEALLCALLS: "TakeAllCalls",
    
    /**
     * value: DoNotAcceptAnyCalls
     * @const
     */
    DONOTACCEPTANYCALLS: "DoNotAcceptAnyCalls",
    
    /**
     * value: DoNotAcceptDepartmentCalls
     * @const
     */
    DONOTACCEPTDEPARTMENTCALLS: "DoNotAcceptDepartmentCalls",
    
    /**
     * value: TakeDepartmentCallsOnly
     * @const
     */
    TAKEDEPARTMENTCALLSONLY: "TakeDepartmentCallsOnly"
  };  /**
   * Allowed values for the <code>presenceStatus</code> property.
   * @enum {String}
   * @readonly
   */
  exports.PresenceStatusEnum = { 
    /**
     * value: Offline
     * @const
     */
    OFFLINE: "Offline",
    
    /**
     * value: Busy
     * @const
     */
    BUSY: "Busy",
    
    /**
     * value: Available
     * @const
     */
    AVAILABLE: "Available"
  };  /**
   * Allowed values for the <code>telephonyStatus</code> property.
   * @enum {String}
   * @readonly
   */
  exports.TelephonyStatusEnum = { 
    /**
     * value: NoCall
     * @const
     */
    NOCALL: "NoCall",
    
    /**
     * value: CallConnected
     * @const
     */
    CALLCONNECTED: "CallConnected",
    
    /**
     * value: Ringing
     * @const
     */
    RINGING: "Ringing",
    
    /**
     * value: OnHold
     * @const
     */
    ONHOLD: "OnHold",
    
    /**
     * value: ParkedCall
     * @const
     */
    PARKEDCALL: "ParkedCall"
  };  /**
   * Allowed values for the <code>userStatus</code> property.
   * @enum {String}
   * @readonly
   */
  exports.UserStatusEnum = { 
    /**
     * value: Offline
     * @const
     */
    OFFLINE: "Offline",
    
    /**
     * value: Busy
     * @const
     */
    BUSY: "Busy",
    
    /**
     * value: Available
     * @const
     */
    AVAILABLE: "Available"
  };

  return exports;
}));
